check/src/unify_type.rs

```diff
@@ -180,16 +182,22 @@ fn do_zip_match<'a, U>(self_: &ArcType,
                 }
                 Less => {
                     let offset = r_args.len() - l_args.len();
-                    let new_type =
-                        unifier.try_match(l, &Type::app(r.clone(), r_args[..offset].into()));
+
+                    let reduced_r = Type::app(r.clone(),
+                                              r_args[..offset].iter().cloned().collect());
```

Comment:
I find it hard to read with to many superfluous intermediates as well (neither of the names `reduced_r` or `reduced_r_args` says anything). I will see what I can do, may be able to simply the code overall.